sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadViaSDF.java

```diff
@@ -0,0 +1,861 @@
+          startOffset =
+              consumerSpEL.offsetForTime(
+                  offsetConsumer,
+                  kafkaSourceDescription.getTopicPartition(),
+                  kafkaSourceDescription.getStartReadTime());
+        } else {
+          startOffset = offsetConsumer.position(kafkaSourceDescription.getTopicPartition());
+        }
+        return new OffsetRange(startOffset, Long.MAX_VALUE);
+      }
+    }
+
+    @GetInitialWatermarkEstimatorState
+    public Instant getInitialWatermarkEstimatorState(@Timestamp Instant currentElementTimestamp) {
+      return currentElementTimestamp;
+    }
+
+    @NewWatermarkEstimator
+    public WatermarkEstimatorT newWatermarkEstimator(
+        @WatermarkEstimatorState Instant watermarkEstimatorState) {
+      return createWatermarkEstimatorFn.apply(watermarkEstimatorState);
+    }
+
+    @GetSize
+    public double getSize(
+        @Element KafkaSourceDescription kafkaSourceDescription,
+        @Restriction OffsetRange offsetRange)
+        throws Exception {
+      double numOfRecords =
+          ((HasProgress) restrictionTracker(kafkaSourceDescription, offsetRange))
+              .getProgress()
+              .getWorkRemaining();
+
+      // Before processing elements, we don't have a good estimated size of records and offset gap.
+      if (avgOffsetGap.containsKey(kafkaSourceDescription.getTopicPartition())) {
+        numOfRecords =
+            numOfRecords / (1 + avgOffsetGap.get(kafkaSourceDescription.getTopicPartition()).get());
+      }
+      return (!avgRecordSize.containsKey(kafkaSourceDescription.getTopicPartition())
+              ? 1
+              : avgRecordSize.get(kafkaSourceDescription.getTopicPartition()).get())
+          * numOfRecords;
+    }
+
+    @NewTracker
+    public RestrictionTracker<OffsetRange, Long> restrictionTracker(
+        @Element KafkaSourceDescription kafkaSourceDescription,
+        @Restriction OffsetRange restriction) {
+      Map<String, Object> updatedConsumerConfig =
+          overrideBootstrapServersConfig(consumerConfig, kafkaSourceDescription);
+      KafkaLatestOffsetEstimator offsetPoller =
+          new KafkaLatestOffsetEstimator(
+              consumerFactoryFn.apply(
+                  KafkaIOUtils.getOffsetConsumerConfig(
+                      "tracker-" + kafkaSourceDescription.getTopicPartition(),
+                      offsetConsumerConfig,
+                      updatedConsumerConfig)),
+              kafkaSourceDescription.getTopicPartition());
+      return new GrowableOffsetRangeTracker(restriction.getFrom(), offsetPoller);
+    }
+
+    @ProcessElement
+    public ProcessContinuation processElement(
+        @Element KafkaSourceDescription kafkaSourceDescription,
+        RestrictionTracker<OffsetRange, Long> tracker,
+        WatermarkEstimator watermarkEstimator,
+        OutputReceiver<KafkaRecord<K, V>> receiver) {
+      // If there is no future work, resume with max timeout and move to the next element.
```

Comment:
```suggestion
      // If there is no known work, resume with max timeout and move to the next element.
```

Reply:
I know I suggested this but am wary now. I think it won't be as performant but checking to see if there are any messages seems like the safer bet (effectively the logic you had before).